### PR TITLE
KAFKA-16427: KafkaConsumer#position() does not respect timeout when group protocol is CONSUMER

### PR DESCRIPTION
**Pull request by kirktrue**
The AsyncKafkaConsumer implementation of `position(TopicPartition, Duration)` was not updating its internal `Timer`, causing it to execute the loop forever. Adding a call to update the `Timer` at the bottom of the loop fixes the issue.

An integration test was added to catch this case; it fails without the newly added call to `Timer.update(long)`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -906,6 +906,7 @@ public long position(TopicPartition partition, Duration timeout) {
                     return position.offset;
 
                 updateFetchPositions(timer);
+                timer.update();
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:909]: (this comment is unrelated to this PR)

It seems `AsyncConsumer#position` does not honour `WakupException`? see following test
```scala
  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
  @Timeout(10)
  def testPositionRespectsWakeup(quorum: String, groupProtocol: String): Unit = {
    val topicPartition = new TopicPartition("abc", 15)
    val consumer = createConsumer()
    consumer.assign(List(topicPartition).asJava)
    val service = Executors.newSingleThreadExecutor()
    service.execute(() => {
      TimeUnit.SECONDS.sleep(1)
      consumer.wakeup()
    })
    try assertThrows(classOf[WakeupException], () => consumer.position(topicPartition, Duration.ofSeconds(3)))
    finally {
      service.shutdownNow()
      service.awaitTermination(1, TimeUnit.SECONDS)
    }
  }
```

**Review by chia7712 — APPROVED** [review]
@kirktrue nice fix. I leave a comment about another possible bug. If that is true, please feel free to fix it in another PR

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 1 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:909]: BTW, please let's me know if you have no free time. I'm fine to fix it if above bug I described is existent

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerCommitTest.scala [hunk 2]
```diff
@@ -271,6 +272,19 @@ class PlaintextConsumerCommitTest extends AbstractConsumerTest {
     consumer.commitSync()
   }
 
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
+  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
+  @Timeout(15)
+  def testPositionRespectsTimeout(quorum: String, groupProtocol: String): Unit = {
```
Comment (lianetm) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerCommitTest.scala]: This test covers a position API behaviour that is unrelated to the committed offsets as I see it (we expect exactly the same when retrieving positions based on the partitions offsets from the leader instead of the committed offsets from the coordinator), so I would say it shouldn't be in this file ? Since there is not specific file for positions-related tests I would suggest we move it to the generic `PlainTextConsumerTest`. Makes sense?

Comment (kirktrue) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerCommitTest.scala]: > Since there is not specific file for positions-related tests I would suggest we move it to the generic `PlainTextConsumerTest`. Makes sense?

I agree. Before I stuck it there, I looked around for any other tests named `test*Position*`. The only test I saw was `testPositionAndCommit` in `PlaintextConsumerCommitTest`, so 🤷‍♂️

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:909]: Thanks @chia7712 for catching that! Can you file a separate Jira to track? Thanks!

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerCommitTest.scala (hunk 2 above)
Comment (kirktrue) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerCommitTest.scala]: Done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 1 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:909]: https://issues.apache.org/jira/browse/KAFKA-16659

**Review by lianetm — APPROVED** [review]
Thanks for the updates @kirktrue. Test failures seem unrelated, LGTM.
